src/jail_rules.c

```diff
@@ -0,0 +1,933 @@
+static
+int
+jail_rules_find_profile(
+    GPtrArray* list,
+    const char* path)
+{
+    int i;
+
+    for (i = 0; i < list->len; i++) {
+        const JailProfile* profile = g_ptr_array_index(list, i);
+
+        if (!strcmp(profile->path, path)) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+static
+JailProfile*
+jail_rules_profile_new(
+    const char* path,
+    gboolean require)
+{
+    /*
+     * Allocate the whole thing from a single memory block, so that a NULL
+     * terminated array of these things can be deallocated with g_strfreev.
+     */
+    gsize len = strlen(path);
+    JailProfile* jail_pro = g_malloc(G_ALIGN8(sizeof(JailPath)) + len + 1);
+    char* ptr = (char*)jail_pro + G_ALIGN8(sizeof(JailPath));
+
+    memcpy(ptr, path, len + 1);
+    jail_pro->path = ptr;
+    jail_pro->require = require;
+    return jail_pro;
+}
+
+static
+void
+jail_rules_add_profile(
+    JailRulesData* data,
+    const JailConf* conf,
+    const char* name,
+    gboolean require)
+{
+    char* path;
+
+    if (g_str_has_suffix(name, SAILJAIL_PROFILE_SUFFIX)) {
+        path = g_build_filename(conf->perm_dir, name, NULL);
+    } else {
+        char* file = g_strconcat(name, SAILJAIL_PROFILE_SUFFIX, NULL);
+
+        path = g_build_filename(conf->perm_dir, file, NULL);
+        g_free(file);
+    }
+
+    if (g_file_test(path, G_FILE_TEST_EXISTS)) {
+        GPtrArray* list = data->profiles;
+        const int prev_index = jail_rules_find_profile(list, path);
+
+        GDEBUG("Adding profile %s%s", path, require ? "" : " (optional)");
+        if (prev_index >= 0) {
+            JailProfile* prev = g_ptr_array_index(list, prev_index);
+
+            /* If the previous one was required, this one must be too */
+            if (prev->require) require = TRUE;
+
+            /* Remove the previous occurrence and complain about it */
+            GWARN("Profile %s is specified more than once", path);
+            g_ptr_array_remove_index(list, prev_index);
+        }
+        g_ptr_array_add(list, jail_rules_profile_new(path, require));
+    } else {
+        GWARN("Profile %s doesn't exist", path);
+    }
+    g_free(path);
+}
+
+static
+int
+jail_rules_find_path(
+    GPtrArray* list,
+    const char* path)
+{
+    int i;
+
+    for (i = 0; i < list->len; i++) {
+        const JailPath* jp = g_ptr_array_index(list, i);
+
+        if (!strcmp(jp->path, path)) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+static
+JailPath*
+jail_rules_path_new(
+    const char* path,
+    gboolean allow,
+    gboolean require)
+{
+    /*
+     * Allocate the whole thing from a single memory block, so that a NULL
+     * terminated array of these things can be deallocated with g_strfreev.
+     */
+    gsize len = strlen(path);
+    JailPath* jail_path = g_malloc(G_ALIGN8(sizeof(*jail_path) + len + 1));
+    char* dest = (char*)jail_path + G_ALIGN8(sizeof(*jail_path));
+
+    memcpy(dest, path, len + 1);
+    jail_path->path = dest;
```

Comment:
so, we can have midflight conflicts in PRs too... just noting that having "char path[];" at end of struct would a) simplify this a bit and b) not hide where/how the data is stored


    JailPath* jail_path = g_malloc(*jail_path + len + 1);
    strcpy(jail_path->path, path);

Reply:
Since `JailPath` is part of the plugin API, I would like it keep it extendible. Adding a field of a variable size to the end of the structure would make that impossible.